receiver/receivercreator/hints.go

```diff
@@ -0,0 +1,223 @@
+)
+
+const (
+	otelHints            = "opentelemetry.io.discovery"
+	discoveryEnabledHint = "enabled"
+	scraperHint          = "scraper"
+	signalsHint          = "signals"
+	configHint           = "config."
+)
+
+// HintsTemplatesBuilder creates configuration templates from provided hints.
+type HintsTemplatesBuilder interface {
+	createScraperTemplateFromHints() ([]receiverTemplate, error)
+}
+
+// K8sHintsBuilder creates configurations from hints provided as Pod's annotations.
+type K8sHintsBuilder struct {
+	logger *zap.Logger
+	config K8sHintsConfig
+}
+
+// createScraperTemplateFromHints creates a receiver configuration based on the provided hints.
+// Hints are extracted from Pod's annotations.
+// Scraper configurations are only created for Port Endpoints.
+// TODO: container_logs configurations are only created for Pod Container Endpoints.
+func (builder *K8sHintsBuilder) createScraperTemplateFromHints(env observer.EndpointEnv) (*receiverTemplate, error) {
+	var endpointType string
+	var podUID string
+	var annotations map[string]string
+
+	endpointType = getStringEnv(env, "type")
+	if endpointType == "" {
+		return nil, fmt.Errorf("could not get endpoint type: %v", zap.Any("env", env))
+	}
+
+	if endpointType != string(observer.PortType) {
+		return nil, nil
+	}
+
+	builder.logger.Debug("handling hints for added endpoint", zap.Any("env", env))
+
+	if pod, ok := env["pod"]; ok {
+		endpointPod, ok := pod.(observer.EndpointEnv)
+		if !ok {
+			return nil, fmt.Errorf("could not extract endpoint's pod: %v", zap.Any("endpointPod", pod))
+		}
+		ann := endpointPod["annotations"]
+		if ann != nil {
+			annotations, ok = ann.(map[string]string)
+			if !ok {
+				return nil, fmt.Errorf("could not extract annotations: %v", zap.Any("annotations", ann))
+			}
+			if len(annotations) == 0 {
+				return nil, nil
+			}
+		}
+		podUID = endpointPod["uid"].(string)
+	} else {
+		return nil, nil
+	}
+
+	if builder.config.Enabled {
+		// Only handle Endpoints of type port for scraper
+		return builder.createScraper(annotations, env, podUID)
+	}
+	return nil, nil
+}
+
+func (builder *K8sHintsBuilder) createScraper(
+	annotations map[string]string,
+	env observer.EndpointEnv,
+	podUID string) (*receiverTemplate, error) {
+
+	var port uint16
+	portName := getStringEnv(env, "name")
+
+	if !discoveryEnabled(annotations, portName) {
+		return nil, nil
+	}
+
+	subreceiverKey := getHintAnnotation(annotations, scraperHint, portName)
+	if subreceiverKey == "" {
+		// no scraper hint detected
+		return nil, nil
+	}
+	builder.logger.Debug("handling added hinted receiver", zap.Any("subreceiverKey", subreceiverKey))
+
+	defaultEndpoint := getStringEnv(env, "endpoint")
+	userConfMap := getConfFromAnnotations(annotations, defaultEndpoint, portName)
+
+	signalsAnn := getHintAnnotation(annotations, signalsHint, portName)
+	signals := getSignalsConf(signalsAnn)
+	userConfMap["signals"] = signals
```

Comment:
Wdyt if we extend the userConfMap to a struct with the unparsed raw configuration of the receiver `map[string]any` and the actual known fields like signal?

I think this would prevent some type casts and deleting unused fields: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35617/files#diff-b4ba0783a3cc06542046d42bf63024247eba48e06e186b87dad0b0a1ed189e95R102

Reply:
Actually, I made `signals` a field of the [`receiverTemplate`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/0898be1e204e10dc8127699743f6bf9a62a529ed/receiver/receivercreator/config.go#L55) since it has similar "weight" with the `rule` field. In the future we could even expose this as a setting to the users so as it can be applied to static receiver_creator's configs. Let me know what you think.

Reply:
That looks good, thanks :+1: